pandas/util/_exceptions.py

```diff
@@ -0,0 +1,16 @@
+import contextlib
+
+
+@contextlib.contextmanager
+def rewrite_exception(old_name, new_name):
```

Comment:
what is this for?

Reply:
This is being used to overwrite exception messages so they display the correct class name.  Looks like it's currently just used for `IntervalIndex` when it calls an underlying `IntervalArray` method that could raise some with an `IntervalArray` specific message, e.g. `'category, object, and string subtypes are not supported for IntervalArray'`.

I suppose we could try to either make these messages slightly more generic ("IntervalArray" --> "IntervalArray-like" or "Interval classes" maybe) or try implementing an additional common interface.

Reply:
What about something like defining a custom exception type in arrays/interval.py and then

```
def _raise_exception(self):
    # We have access to the correct `type(self).__name__`
    pass

def method(self):
    try:
        do_thing()
    except _IntervalException:
        self._raise_exception()
```

does that make sense?

Reply:
Or we could just use IntervalArray in the exception message. Doesn't really matter to me.

Reply:
Yeah, I don't really have a preference either.  I might end up trying a few things to see what's cleanest.